File: solutions/haskell/127-meeting-rooms-ii.hs
{-
  Problem 127: Meeting Rooms II (LeetCode 253)
  Difficulty: Med
  Language: Haskell
-}
import Data.List (sort)

minMeetingRooms :: [(Int, Int)] -> Int
minMeetingRooms intervals =
  let starts = sort (map fst intervals)
      ends = sort (map snd intervals)
      go [] _ rooms maxR = maxR
      go (s:ss) (e:es) rooms maxR
        | s < e = go ss (e:es) (rooms + 1) (max maxR (rooms + 1))
        | otherwise = go ss es rooms maxR
  in go starts ends 0 0

File: solutions/typescript/127-meeting-rooms-ii.ts
/**
 * Problem 127: Meeting Rooms II (LeetCode 253)
 * Difficulty: Med
 * Language: TypeScript
 */
function minMeetingRooms(intervals: number[][]): number {
  const starts = [...intervals].map(i => i[0]).sort((a, b) => a - b);
  const ends = [...intervals].map(i => i[1]).sort((a, b) => a - b);
  const go = (si: number, ei: number, rooms: number, maxR: number): number => {
    if (si >= starts.length) return maxR;
    if (starts[si] < ends[ei])
      return go(si + 1, ei, rooms + 1, Math.max(maxR, rooms + 1));
    return go(si + 1, ei + 1, rooms, maxR);
  };
  return go(0, 0, 0, 0);
}
